Hybrid newton flag

lechevaa opened the pull request:
I'm a Postdoc at NORCE, this is my first PR to OPM. 

Added Hybrid Newton Nonlinear Preconditioning Flags and Features for three phase compositional black oil problem. 

Reference to the introduction of the method within OPM: 
Lechevallier, Antoine, Sandve, Tor Harald, Landa-Marbán, David, Kane, Birane, and Sarah Eileen Gasda. "Incremental Machine Learning for Near-Well Prediction: A Non-Linear Preconditioning Approach to Faster History Matching." Paper presented at the SPE Reservoir Simulation Conference, Galveston, Texas, USA, March 2025. doi: https://doi.org/10.2118/223843-MS

Short description about HybridNewton method:
The goal is simply to update the initial guess of Newton's method with a new value closer to the solution to speed up the convergence. The update is done using a neural network. 

What is new compared to the paper: 
I finally managed to fix the Pressure prediction ! This is now fully supported

Some general information about the implementation:

- The user provides a json file with all required information i.e path to neural network, input features, output features, time to apply the model, cell indexes, feature engineering, scalers. 

- Multiple models can be provided within a single config file. 

- The units are assumed to be the same as what is outputted by opm. 

- For now I only implemented for FlowProblemBlackOil. 

Input features supported:
- Dynamic: PRESSURE, SWAT, SGAS, SOIL, RS, RV (same name as what is outputted by opm)
- Static: TIMESTEP (scalar feature!), PERMX

Output Features supported: 
- absolute dynamic: PRESSURE, SWAT, SGAS, SOIL, RS, RV 
- relative dynamic:  DELTA_PRESSURE, DELTA_SWAT, DELTA_SGAS, DELTA_SOIL, DELTA_RS, DELTA_RV 

New Flags: 
--use-hy-ne=bool : use or not the HybridNewton method 
--hy-ne-config-file=path : Path to the configuration file (json) that allows OPM to interpret correctly how to precondition Newton's method. 

New files:

opm/simulators/flow/HybridNewtonConfig.hpp

struct HybridNewtonConfig that parses the main information required to run HybridNewton

opm/simulators/flow/HybridNewton.hpp

Contains the class implementation of BlackOilHybridNewton. The main method is runHybridNewton given a HybridNewtonConfig.  

What is missing for now:
I created a python code that tests all supported features in all combinations possible and generate neural network models together with config files.

I have done some tests in python, running Drogon and SPE1 and predicting the exact solution to ensure that Newton's method converges in 0 iteration. 

However these tests are not yet included within the opm python testing framework. I need to be careful about dependencies (same as the one in opm-common/python/opm/ml/ml_tools/requirements.txt) and testing. 

TODO:
adapt my existing tests to the opm-simulator python testing

akva2 commented on opm/simulators/flow/FlowProblemParameters.hpp:
```diff
@@ -56,6 +56,11 @@ struct OutputMode { static constexpr auto value = "all"; };
 struct RestartWritingInterval { static constexpr int value = 0xffffff; }; // disable
 
+// Path to the config file containing all Hybrid Newton parameters
+struct HyNeConfigFile { static constexpr auto value = "foo.json"; }; 
```
please choose a better default

akva2 commented on opm/simulators/flow/HybridNewton.hpp:
```diff
@@ -0,0 +1,648 @@
+    {
+        // Load cell indices from file at runtime
+        std::vector<int> cell_indices = loadCellIndicesFromFile(config.cell_indices_file);
+        const size_t n_cells = cell_indices.size();
```
use std::size_t

akva2 commented on opm/simulators/flow/HybridNewton.hpp:
```diff
@@ -0,0 +1,648 @@
+
+
+protected:
+
```
drop newline

akva2 commented on opm/simulators/flow/HybridNewton.hpp:
```diff
@@ -0,0 +1,648 @@
+        }
+    }
+
+    std::vector<HybridNewtonConfig> getApplicableModels(double current_time)
```
non-const getter ?

akva2 commented on opm/simulators/flow/HybridNewton.hpp:
```diff
@@ -0,0 +1,648 @@
+        return applicable;
+    }
+
+    bool shouldApplyHybridNewton(double current_time, const HybridNewtonConfig& config)
```
const?

akva2 commented on opm/simulators/flow/HybridNewton.hpp:
```diff
@@ -0,0 +1,648 @@
+    }
+
+    
+    void validateAllConfigs()
```
smells like a const

akva2 commented on opm/simulators/flow/HybridNewton.hpp:
```diff
@@ -0,0 +1,648 @@
+    * \return A tensor of shape [n_cells x n_input_features], where rows
+    *         correspond to cells and columns correspond to input features.
+    */
+    Opm::ML::Tensor<Evaluation> 
```
You are already in the Opm namespace, avoid Opm:: where possible.

akva2 commented on opm/simulators/flow/HybridNewton.hpp:
```diff
@@ -0,0 +1,648 @@
+    * \throws std::runtime_error if an expected output feature is missing
+    *         or if state update fails for any cell.
+    */
+    void updateInitialGuess(Opm::ML::Tensor<Evaluation>& output, const HybridNewtonConfig& config, const std::vector<int>& cell_indices, size_t n_cells)
```
break long line

akva2 commented on opm/simulators/flow/HybridNewtonConfig.hpp:
```diff
@@ -0,0 +1,135 @@
+} // namespace Opm
+
+
+#endif
```
eol at eof

akva2 commented:
jenkins build this please

daavid00 commented:
jenkins build this please

daavid00 commented:
jenkins build this please

lechevaa commented:
Thanks again for your reviews ! 

I tried to refactor the two files as you mentioned. I hope the readability is much better now. 

I also added some Python tests.
Regarding them, I am using only the opm.io.ecl module for managing data. 

I wanted to also use the from opm.simulators.BlackOilSimulator, however I ran into some issues using it. 
The first issue is that it is not possible to instantiate two BlackOilSimulator objects, making comparison between a baseline simulation and a Hybrid Newton simulation not directly possible. 
To overcome this, I thought that I could have a single instance of BlackOilSimulator, and modify the fluid state to the previous timestep solution so that I can run it again using HybridNewton. This is also not supported for now, or I could not find it. 

If one of this functionality becomes available, I can rewrite the code using more opm python parts and avoid using subprocess to call flow. 

Finally, I struggled a bit to set up the imports, so I am not convinced these python tests will go through CI due to import errors.

lechevaa commented:
> Thank you for the changes. Several parts were easier for me to understand now.
> 
> If you manage to get some tests in before the release that would be good but it is not absolutely necessary, we certainly don't have a perfect code coverage already.
> 
> Good that you are thinking of adding a tutorial. I think it would make sense to create a repo directly into the OPM project then, as this feature would depend on that tutorial.
> 
> One general comment to consider is that I see you have quite extensively dropped using brackets. This will make the code more concise but can hurt readability and maintainability, and easily lead to introducing bugs when adding new things later. So I would consider to avoid using this format. If you are not convinced you can get some more opinions here https://stackoverflow.com/questions/12193170/whats-the-purpose-of-using-braces-i-e-for-a-single-line-if-or-loop.

Sorry in the previous commit, I refactored the code and forgot some brackets at the specific place you pointed out, it was not on purpose. I updated it back so there shall be brackets everywhere now. If I am still missing some, I apologize and will update it.

jakobtorben commented:
Thanks for all the changes. I think it is close to being ready now. Go over the docs in `HybridNewton.hpp` and `HybridNewtonConfig.hpp` to make sure that they are updated given recent changes, and check the formatting in `HybridNewtonConfig.hpp`. The test pipeline also needs to be fixed before we can merge it in, either with or without tests.

lechevaa commented:
The test pipeline is fixed and I updated the documentation for all the important functions. 
Do not hesitate to tell me if there is anything else to improve before it to be ready, and thanks again for the time you spend on reviewing !
